backfiller/backfiller/main.py

```diff
@@ -129,159 +131,263 @@ def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
 
-	backfill(base_dir, stream, variants, nodes=nodes)
-	
-	# I'm sure there is a module that does this in a more robust way 
-	# but I understand this and it gives the behaviour I want
-	while True:
 
-		now = datetime.datetime.now()
+class BackfillerWorker(object):
+	"""Backfills segments from a node.
 
-		if now - full_fill_start > datetime.timedelta(minutes=full_fill_wait):
+	Backfills all segments from node/stream to base_dir/stream for all variants.
+	Every SMALL_INTERVAL minutes backfill the last three hours starting from the
+	most recent one (a 'small backfill'). When not doing a small backfill,
+	backfill all segments starting with the most recent one (a 'large backfill')
+	unless a large backfill has occured less than LARGE_INTERVAL ago.""" 
 
-			backfill(base_dir, stream, variants, nodes=nodes)
+	SMALL_INTERVAL = 5 #minutes between small backfills
+	LARGE_INTERVAL = 60 #minutes between large backfills
+	WAIT_INTERVAL = 1 #seconds between backfill actions
 
-			fill_start = now
-			full_fill_start = fill_start
+	def __init__(self, manager, base_dir, node, stream, variants):
+		"""Constructor for BackfillerWorker"""
+		self.manager = manager
+		self.logger = manager.logger.getChild('BackfillerWorker({})'.format(node))
+		self.base_dir = base_dir
+		self.node = node
+		self.stream = stream
+		self.variants = variants
+		self.stopping = gevent.event.Event()
+		self.done = gevent.event.Event()
 
-		elif now - fill_start > datetime.timedelta(minutes=fill_wait):
 
-			backfill(base_dir, stream, variants, 3, nodes=nodes)
+	def __repr__(self):
+			return '<{} at 0x{:x} for {!r}/{!r}>'.format(type(self).__name__, id(self), self.node, self.stream)
+	__str__ = __repr__
 
-			fill_start = now
 
+	def stop(self):
+		"""Tell the worker to shut down"""
+		self.logger.info('Stopping')
+		self.stopping.set()
+
+
+	def backfill(self, hours, segment_order='random', recent_cutoff=60):
+		"""Backfill from remote node.
+	
+		Backfill from node/stream/variants to base_dir/stream/variants for each
+		hour in hours.
+	
+		Keyword arguments:
+		segment_order -- If 'random', randomise the order of segments (default).
+			If 'forward', sort the segment in ascending order. If 'reverse',
+			sort the segments in descending order. Otherwise, do not change the
+			order of segments.
+		recent_cutoff -- Skip backfilling segments younger than this number of
+			seconds to prioritise letting the downloader grab these segments."""
+	
+		#logging.info('Starting backfilling from {}'.format(node))
+	
+		for variant in self.variants:
+	
+			for hour in hours:
+	
+				self.logger.info('Backfilling {}/{}'.format(variant, hour))
+	
+				local_segments = set(list_local_segments(self.base_dir, self.stream, variant, hour))
+				remote_segments = set(list_remote_segments(self.node, self.stream, variant, hour))
+				missing_segments = list(remote_segments - local_segments)
+	
+				# useful if running in parallel so multiple nodes don't request the same segment at the same time
```

Comment:
The idea with the random order is to lessen the odds that two workers request the same segment at the same time. I think it is more likely that one node will be requesting an hour from a multiple nodes at the same time (for example on start up) than multiple nodes requesting the same hour from a single node at the same. If multiple workers are trying to backfill the same hour at the same time and the segments are requested in order, the workers will request the same segment at the same time, leading to duplicated effort.

Reply:
Hm. I prefer requesting things in order because it makes it clearer to the operator what the backfiller is doing / where it's up to. But not downloading the same thing twice is also a nice property. I guess just update the comment to make it clearer?